### DISCUSSION TITLE
Revisit to Kiali Governance

### DISCUSSION THREAD

**Discussion by israel-hdez** [kiali/kiali#4199] [Governance]
I would like to start a discussion to review and update Kiali's project governance which hasn't been reviewed for about half a year.

This is motivated because of the growing number of reported issues from community, and also because of interest in contributing to the project. Secondarily, there is also some motivation because of [the new larger efforts being proposed](https://github.com/kiali/kiali/discussions/categories/roadmap).

Although "governance" is sometimes seen as something bad and bureaucratic, [a well defined governance helps in project engagement and more contributions](https://www.theopensourceway.org/the_open_source_way-guidebook-2.0.html#_why_governance).

### References to understand governance 

Some references for understanding what's governance and how to define it:
[1] https://opensource.com/article/20/5/open-source-governance
[2] https://opensource.guide/leadership-and-governance/#what-are-some-of-the-common-governance-structures-for-open-source-projects
[3] http://oss-watch.ac.uk/resources/governancemodels#how-do-open-development-projects-make-decisions
[4] https://www.theopensourceway.org/the_open_source_way-guidebook-2.0.html#_project_and_community_governance

IMO, [4] is the most detailed guide.

### Existing governance

At the time of writing, Kiali project provides the following governance document: https://github.com/kiali/kiali/blob/3b248fe330f7eb05919c9680f8663de2b25110f4/GOVERNANCE.md. It describes the _maintainer_ role and how to gain that status. 

But on the cited references, the following documents can also be considered as part of governance:
* Code of conduct: https://github.com/kiali/kiali/blob/3b248fe330f7eb05919c9680f8663de2b25110f4/CODE_OF_CONDUCT.md
* Contribution process: https://github.com/kiali/kiali/blob/3b248fe330f7eb05919c9680f8663de2b25110f4/CONTRIBUTING.md and https://kiali.io/contribute/
* Security reporting: https://github.com/kiali/kiali/blob/3b248fe330f7eb05919c9680f8663de2b25110f4/SECURITY.md
* Release process: https://github.com/kiali/kiali/blob/3b248fe330f7eb05919c9680f8663de2b25110f4/RELEASING.adoc
* Coding styles: https://github.com/kiali/kiali/blob/3b248fe330f7eb05919c9680f8663de2b25110f4/STYLE_GUIDE.adoc and https://github.com/kiali/kiali-ui/blob/e612ba03baee852257e82f2e505dfe645a756b0d/STYLE_GUIDE.adoc
* How to communicate: https://kiali.io/contribute/#_talk_with_us

When comparing all these documents and actual behavior of Kiali project, IMO what is written is what was though to be ideal. Although a lot of it is being followed, some parts need to be updated with current practices.

### Goal of this discussion

My aim **isn't to build something new**. Kiali project already has a well working governance. In reference [4], this is described as an _implicit governance_. 

However a consequence of an implicit governance is that, sometimes, discussions can reach a stall. IMO, this is because expectations and roles are not explicitly set.

So, my aim is to clearly identify this implicit governance and write it down to make it explicit. **Let's identify current member roles, policies, processes and create one or more documents describing all this**.

### Starting the discussion

In the current [GOVERNANCE.md](https://github.com/kiali/kiali/blob/3b248fe330f7eb05919c9680f8663de2b25110f4/GOVERNANCE.md) document, we only have the concept of _maintainer_, while [4] mentions three basic roles: members, contributors and leaders. Although IMO in Kiali we have these three basic roles, I would like to discuss in a thread if we do, or if we have less or more roles, what are the current responsibilities and what's needed to step in, or step down each role. For example, it comes into my mind that we may also have the additional _maintainer_ role, which happens to be the only one described in our GOVERNANCE.md file. Probably, we need to identify overall project roles and scoped roles. For example, I can identify that @jmazzitelli is a contributor (and maintainer?) of Kiali project and I can also identify that he is leading the kiali-operator part.

It's worth to emphasize that the goal is to identify and make explicit current Kiali governance and NOT to build anything new. This part of "roles" and "policies and procedures" is the _basic_ governance and Kiali does not have it documented right. So, I hope we can clearly identify this.

For other parts, I would like to discuss them in separate threads (under this same discussion):
* Code of conduct: I see outdated contacts at the end of the document. We should discuss what should be the contacts.
* Contribution process: I see outdated information. Let's discuss what to update.
* Security reporting: IMO looks good. Anyway, I encourage people to review it and raise any concerns with it.
* Release process: I encourage people to review it and tell if it's OK. In [4], there is this section about the release process: https://www.theopensourceway.org/the_open_source_way-guidebook-2.0.html#_release_process. Given this reference, should we document anything else?
* Coding styles: Honestly, is anybody really following coding styles? or "in linters we trust"?
* How to communicate: IMO, this also needs some attention. Sometimes, people trying to reach us is bounced between the different contact channels. Let's identify how we are using the different contact channels and write this down, so that new people uses the right channel since the beginning and prevent bouncing.

And finally, is there something else?

**Comment by abonas** [comment]
Thanks for starting this super important discussion.
I'd like to add some points that IMO need to be either refined or added to the governance:
1. When someone should stop being a maintainer. https://github.com/kiali/kiali/blob/3b248fe330f7eb05919c9680f8663de2b25110f4/GOVERNANCE.md#changes-in-maintainership
At the moment this is either by resignation or by voting of maintainers. It is not stated under which circumstances such a vote will occur. 
I suggest adding some criteria when someone should stop being a maintainer based on lack of involvement in the project (as an example - no code contribution/code review in X months). Then when this criteria is met, a vote can be initiated.

2. When the governance will start being enforced on all terms. The addition of maintainers has not been fully following the criterias that are outlined in the governance doc, and there is a need to define that starting date X all maintainer addition/removal will be done strictly per governance document.

3. Maintainers vote - need to clarify where/how this vote will happen? During community meeting? On Github discussion offline?

4. "a 2/3 majority vote by maintainers" - what if most of maintainers don't vote on some decision? What number of maintainers will determine the outcome of the vote? is this 2/3 out of those who actually voted?

5. Decision making on features/architecture/direction of the project. Who should have the final say on decisions if there's no consensus? As an example, discussion about whether Kiali should have a layer that acts like federation of p8s or not, OSM/SMI, etc.

**Reply by israel-hdez** [reply]
About (1), I agree! Although, I think we should evaluate contributions in general (not only code).

About (2), I agree that we are not fully following what is documented. My thought is that we should have followed what we established since day 1. I'm not sure if we can really enforce something. Perhaps it's better to write how the project is currently behaving and remove anything that we are not following??? Although, probably our current behavior is not that good. So, I'm not sure.

About (3), I think a GitHub discussion is our currently best channel to do a voting.

About (4), that's an interesting question. My take is that it should be 2/3 of all current maintainers, regardless of abstention. Then, simply, no majority == no change. However, I'm not sure if we will constantly fall in the no-majority trap. Kiali doesn't have a lot of people involved. So, one missing vote has a lot of weight.

About (5), I think that's part what I'm trying to do when asking for defining "roles", and also defining "policies and procedures". People on certain roles should take final decisions.

**Reply by abonas** [reply]
> About (1), I agree! Although, I think we should evaluate contributions in general (not only code).

Of course, I just provided some examples.
> 
> About (2), I agree that we are not fully following what is documented. My thought is that we should have followed what we established since day 1. I'm not sure if we can really enforce something. Perhaps it's better to write how the project is currently behaving and remove anything that we are not following??? Although, probably our current behavior is not that good. So, I'm not sure.

I think it is legitimate to say "starting date X the governance is fully followed" and while it does not fix the past, it will be more correct towards the future.
> 
> About (3), I think a GitHub discussion is our currently best channel to do a voting.

Sounds good
> 
> About (4), that's an interesting question. My take is that it should be 2/3 of all current maintainers, regardless of abstention. Then, simply, no majority == no change. However, I'm not sure if we will constantly fall in the no-majority trap. Kiali doesn't have a lot of people involved. So, one missing vote has a lot of weight.
> 
> About (5), I think that's part what I'm trying to do when asking for defining "roles", and also defining "policies and procedures". People on certain roles should take final decisions.

What would you suggest in terms of role that can resolve this?

**Reply by israel-hdez** [reply]
> I think it is legitimate to say "starting date X the governance is fully followed" and while it does not fix the past, it will be more correct towards the future.

I'm OK with this :-)

> > About (5), I think that's part what I'm trying to do when asking for defining "roles", and also defining "policies and procedures". People on certain roles should take final decisions.
> 
> What would you suggest in terms of role that can resolve this?

I think it's closely related to what @nrfox mentioned in the ["Process for large enhancements" discussion](https://github.com/kiali/kiali/discussions/4208). IMO, if there is an owner/lead for certain enhancement/feature, the owner/lead of each enhancement/feature is the person who should do final decisions if there is no consensus. That's a responsibility of a "owner/leader" role of a certain enhancement/feature.

This is why I'm asking to identify "scoped" roles, and gave the example that Mazz seems to be the owner/leader of the kiali-operator, while he probably is just a contributor (not an owner/leader) when we look at the Kiali project in general.

I wouldn't like to see Kiali falling in the ["Benevolent Dictator" model](https://www.theopensourceway.org/the_open_source_way-guidebook-2.0.html#_founder_leader). Although, probably, one general owner/leader is needed for cases where consensus is not reached for enhancements that impact a lot of parts of Kiali.

This is what I think. I'm open to listen for better ways to the (5) point (Decision making on features/architecture/direction of the project.)

**Reply by jberkus** [reply]
You don't need a single person to make decisions if there's not a lazy consensus.  Voting via +1 on PRs/Issues is a common practice.  That is, there doesn't have to be a single "owner" if there are multiple people interested/qualified, and it's better if there's not a single owner.  People go on vacation, leave unexpected, have unreasonable pet peeves, etc.   Having a single owner isn't a *feature*, it's a bug, usually caused by "there's only one person who really knows this code currently contributing".

**Reply by israel-hdez** [reply]
@jberkus Thanks for your feedback. Very appreciated! I have a few questions:

> You don't need a single person to make decisions if there's not a lazy consensus. Voting via +1 on PRs/Issues is a common practice.

I wonder how you deal with cases where voting is tie?
I'm mainly thinking about owners (or "main" owner for a specific feature) for cases where we need a tiebreaker. But, in general, I would be happy if most of the time decisions can be reached via voting.

> it's a bug, usually caused by "there's only one person who really knows this code currently contributing"

Oh, well... I think Kiali project is falling on this "bug" :-(
We have a small team of about 10 people, and Kiali codebase has grown to the point where the code of some features is known only by one person. I guess we need more community to get involved, but we either have failed to engage people or, simply, most community are on the "users" side.

**Reply by dneary** [reply]
@israel-hdez 

> I wonder how you deal with cases where voting is tie?

I think it's an edge case. But how does it work now? I would bet that specific areas of the project have people who others defer to when they have a strong opinion on a topic.

> Oh, well... I think Kiali project is falling on this "bug" :-(

I think that while this is a situation every project would like to avoid, it's normal! The median open source project has one developer.

> We have a small team of about 10 people, and Kiali codebase has grown to the point where the code of some features is known only by one person. I guess we need more community to get involved, but we either have failed to engage people or, simply, most community are on the "users" side.

A starting point is to identify those areas and people. Once that knowledge is surfaced, that's a great opportunity to bootstrap an apprenticeship program to allocate understudies to maintainers, and start to build more knowledge of those areas among other project developers.

**Reply by jberkus** [reply]
> Oh, well... I think Kiali project is falling on this "bug" :-(

It's an extremely common problem, and likely any governance-by-feature-owners we wrote today would result in one owner for each feature.  Let's just not write out the possibility of co-owners in the future.

Alternately, if we're only talking about 10 people ... then why can't we just govern by lazy-consensus/majority vote?  10 people is not a lot of folks to poll, for anything.

**Reply by israel-hdez** [reply]
@dneary 
> 
> > I wonder how you deal with cases where voting is tie?
> 
> I think it's an edge case. But how does it work now? I would bet that specific areas of the project have people who others defer to when they have a strong opinion on a topic.
> 

That's right. Honestly, we rarely do voting. We usually defer decisions/direction to the main contributor of the feature :-)

The problem comes when an enhancement has broad impact. No PR is proposed because we opt first for high level discussions, brainstorming, etc. Some of this efforts are quite productive, but nothing settles down resulting in no progress. Sometimes, we have done voting and reached decisions, but if there is a tie for an enhancement with broad impact, it's hard to break the tie or find alternatives. Then, it has been very hard to evolve the discussions to proposals with everybody involved.

Any ideas here? As much as I dislike the idea, IMO this reflects the need of a global leader/owner.

> > Oh, well... I think Kiali project is falling on this "bug" :-(
> 
> I think that while this is a situation every project would like to avoid, it's normal! The median open source project has one developer.
> 

This is an interesting stat :-)
OK, good to know. Perhaps we shouldn't worry too much about it.

> > We have a small team of about 10 people
> 
> A starting point is to identify those areas and people. Once that knowledge is surfaced, that's a great opportunity to bootstrap an apprenticeship program to allocate understudies to maintainers, and start to build more knowledge of those areas among other project developers.

And this is also an interesting idea. Perhaps we should be doing this regularly as people come and go. I'll see what others think about this.


@jberkus 

> > Oh, well... I think Kiali project is falling on this "bug" :-(
> 
> It's an extremely common problem, and likely any governance-by-feature-owners we wrote today would result in one owner for each feature. Let's just not write out the possibility of co-owners in the future.
> 
> Alternately, if we're only talking about 10 people ... then why can't we just govern by lazy-consensus/majority vote? 10 people is not a lot of folks to poll, for anything.

Perhaps what I wrote about enhancements with broad impact can give you the idea. I mean, most of the time we don't do voting because we defer to others. But it's been hard to progress when something has broad impact.

**Reply by xeviknal** [reply]
> When the governance will start being enforced on all terms. 

IMO, The governance is applied when the document is merged into master. Changes in the governance are enforced the moment the merged button is hit.

> The addition of maintainers has not been fully following the criterias that are outlined in the governance doc, and there is a need to define that starting date X all maintainer addition/removal will be done strictly per governance document.

There are two things: the rules and the application of the rules. We just defined the rules and we completely broke them. Literally not taken into any account. And this is our fault. And we should acknowledge and perhaps amend it.

> Maintainers vote - need to clarify where/how this vote will happen? During community meeting? On Github discussion offline?

I have two ideas:
1. Through a PR changing the maintainers file and vote from there.
2. Open a discussion in this repo as @israel-hdez mentioned.

> "a 2/3 majority vote by maintainers" - what if most of maintainers don't vote on some decision? What number of maintainers will determine the outcome of the vote? is this 2/3 out of those who actually voted?

My take here is 2/3 of the maintainers. Like the voting on a parliament. Absence (no vote) or blank vote are not considered to positive. If people don't vote is not a problem on the rule itself it is a problem of the community. And probably a symptom of something is not alright. Therefore, I'd consider doing other activities and changes to foster commitment and implication.


> Decision making on features/architecture/direction of the project. Who should have the final say on decisions if there's no consensus? As an example, discussion about whether Kiali should have a layer that acts like federation of p8s or not, OSM/SMI, etc.

Interesting and I think this need to be very very clear and transparent. This probably is a big stopper to big contributions (we don't have data though). In the past, we got one company interested in contributing into kiali. This was one of the major concern shared by them.
If we'd split the project in different sub-project I guess that we should have 2 votes for each sub-project affecting the change. We should be very very friend in this though. Perhaps drive big changed in smaller steps.

**Reply by israel-hdez** [reply]
> There are two things: the rules and the application of the rules. We just defined the rules and we completely broke them. Literally not taken into any account. And this is our fault. And we should acknowledge and perhaps amend it.

Well, that reference [4] that I posted in the main comment recommends writing out current behavior. So, that suggests that in the past the rules that were established didn't reflect the reality. So, probably, the main fault was to establish something new, rather than simply write current behavior.

> I have two ideas:
> 
> 1. Through a PR changing the maintainers file and vote from there.
> 2. Open a discussion in this repo as @israel-hdez mentioned.

+1

> Therefore, I'd consider doing other activities and changes to foster commitment and implication.

I would like to hear your thoughts about what kind of activities and changes :-)

**Comment by dneary** [comment]
Sweet! Great to see a resource I helped write (The Open Source Way chapter) be helpful!

The most important part to make explicit IMHO is the one that project participants most frequently overlook - who are the project's leaders? How do you decide whether to accept a patch submission or not? Are there parts of the project where someone is expected to review and sign off before a merge? Are there people who get to merge code without review? How is planning done for future releases in terms of big features and prioritization of smaller features? These kinds of day-to-day interactions are the ones that are hardest for newcomers to the project to pick up on, and are the most important for them to understand.

**Reply by israel-hdez** [reply]
That sounds like documenting our workflow :-)
And yes, a lot of these questions are not explicitly answered :-(

**Reply by israel-hdez** [reply]
> Sweet! Great to see a resource I helped write (The Open Source Way chapter) be helpful!

It's very helpful! Good resource!

**Reply by xeviknal** [reply]
I completely agree with this point. Document the day-to-day interactions and processes.
We should have a very clear section describing the life of a issue/PR from the inception to the integration to the main branch: including priorisation, code review criteria and merge criteria. Including the possible authors in each step.

**Comment by jberkus** [comment]
So, some suggestions and commentary:

In the Open Source Way conception, "Maintainer" is the same as "Leader".   What's undefined in current documentation are other roles, such as Contributor, Reviewer, Subproject Owner, etc.  As a sophisticated example of a full-fledged set of roles, see the [CNCF Contributor Ladder template](https://github.com/cncf/project-template/blob/main/CONTRIBUTOR_LADDER.md). Not saying that you need to adopt that, but it does give examples of how many roles a project could have if it needs them.

If you look at the full template for [maintainer-led govenance](https://github.com/cncf/project-template/blob/main/GOVERNANCE-maintainer.md) that Karen Chu, Dawn Foster and I developed for the CNCF, it handles some of the unhandled things (like where votes take place and how someone gets promoted to maintainer).  However, I'm firmly on the side of "a majority vote of existing maintainers" for any decision that requires a vote, because that means it doesn't *matter* where the vote takes place.  It lets you be flexible, take some votes in meetings, others on GH issues, etc.  All of them require a vote of 51% of the maintainers that there are.

BTW, did you know that your list of maintainers is only available to your admins?  There really ought to be a public list somewhere.

Your current governance has all maintainers in a single pool, which is pretty common for projects where there is a single core repository.  However, some of the comments above seem to suggest that you have the equivalent of subproject or feature owners, and if that's the reality, your governance should reflect it.

**Reply by jberkus** [reply]
Further input: the Maintainer-led-project template in the CNCF is intended for projects where the list of "approvers" and the list of project leaders is the same group.  Sometimes it's not though; with Kubevirt, we created a version where the list of maintainers is a subset of the list of approvers, plus some other folks: https://github.com/kubevirt/community/pull/123/files

**Reply by israel-hdez** [reply]
It's very good templates. A contributor ladder is probably too much for Kiali Project, but I that template is nice to understand better what we could need.

> However, I'm firmly on the side of "a majority vote of existing maintainers" for any decision that requires a vote, because that means it doesn't _matter_ where the vote takes place.

I guess voting should be public, right?
What I don't clearly get is if voting should leave a record? I mean, voting via GitHub will leave the record of comments. However, voting via a call won't leave the record unless the call is recorded and you may give only the result of voting. 

> BTW, did you know that your list of maintainers is only available to your admins? There really ought to be a public list somewhere.

Well, we have this: https://github.com/orgs/kiali/people
But, TBH, that list is outdated and a lot of people have chosen to keep their membership as private. So, I reckon is not enough.

> Your current governance has all maintainers in a single pool, which is pretty common for projects where there is a single core repository. However, some of the comments above seem to suggest that you have the equivalent of subproject or feature owners, and if that's the reality, your governance should reflect it.

Yes. And that's one thing I would like to come up in this GitHub discussion.

**Reply by jberkus** [reply]
> What I don't clearly get is if voting should leave a record? I mean, voting via GitHub will leave the record of comments. However, voting via a call won't leave the record unless the call is recorded and you may give only the result of voting.

That's what meeting minutes are for.  Every public meeting should have public minutes and a note-taker.  Not only does that let you take votes, it also means that folks who didn't attend the meeting (say, because of timezones) can keep up on what happened there.

**Reply by jberkus** [reply]
Also -- maybe the full contributor ladder is too much for Kiali, but you should probably have *some kind* of contributor ladder if you're looking to attract non-Hatter contributors.  Right now you have no definitions of any role other than "maintainer", which kinda suggests to new people that there are no other roles.

**Reply by israel-hdez** [reply]
> That's what meeting minutes are for. Every public meeting should have public minutes and a note-taker.

Makes sense. So, the call doesn't need to be recorded and a document with the outcomes is enough.
We currently don't do meetings notes. It's a point to improve.

> Also -- maybe the full contributor ladder is too much for Kiali

That's right. It's too much. And I also agree that having "maintainer" as the only role is not enough.
We have had contributors that never became maintainers.

**Reply by xeviknal** [reply]
wrt the ladder, I think we already have one but it's the smallest expression of it and pretty implicit: contributor to maintainer. After the possible changes in the roles as a result of this discussion, we should keep and adapt the ladder as well.

+1 to say that this ladder is crucial to attract new long-lasting contributors. However, I think it is as important as generating a clear, transparent and trustworthy workflow for merging new code and large contributions. Otherwise, the ladder would be really exhaustive but tough for contributors to climb it.

wrt roles, I think we should reflect more what we already have into the governance. We do have a model close to the one of owners by areas/subprojects/features. So +1 to keep evolving this model and make it more explicit for both of current contributors and future ones.
This is an example from Istio: https://github.com/istio/istio/blob/master/CODEOWNERS I think it is too much. In our case, we can come to a good feature split and assign it to owners.

I like the idea of making the distinction of owners and maintainers.

I think that the maintainers list is this one: https://github.com/orgs/kiali/teams/maintainers/members
I think that this can be fixed now. It looks like a unknown situation than an intentional one. @abonas @jshaughn @lucasponce are you able to make it open to the world and adjust to the current maintainer setup?

**Reply by lucasponce** [reply]
Apologizes if I missed something, just trying to catchup with the conversation.

We have two roles: contributors and maintainers. Being maintainers the ones that have privileges to merge work and participate in governance.

I think those two groups could be enough for a size project as Kiali. It's true that we have followed some implicit processes non documented that would be worth to put in wording.

The CODEOWNERS is interesting, but again, Kiali is not as big enough to have such separation. In today's status, any change would touch a lot of layers (from frontend to backend + operator + doc) so defining owners for those projects sounds a little bit overkill.

Note that one thing that is not commented but is implicitly used in the project since the beginning: when an issue is assigned, it has an owner, and the owner defines the reviewers and QE, and in case of conflicts, the team leads of the project help to solve any point making sure that a particular work is going to integrate well with other on-going effort (giving all visibility in issues and PRs to allow anyone to participate and contribute).

**Reply by israel-hdez** [reply]
@xeviknal 
> wrt the ladder, I think we already have one but it's the smallest expression of it and pretty implicit: contributor to maintainer.
> wrt roles, I think we should reflect more what we already have into the governance.

If I understand correctly, the ladder lists the roles. The ladder is, simply, placing some organization and formalization to roles. So, let's not think on "the ladder" and "roles" as two separate things.

@lucasponce 
> We have two roles: contributors and maintainers. Being maintainers the ones that have privileges to merge work and participate in governance.

Well, I don't think "contributors" and "maintainers" is enough.

Later, you are also mentioning "issue owners", "reviewers", "QE" and "team leads". All these are also roles that people involved in Kiali are performing. We shouldn't be hiding these figures, and we should transparently provide step-in criteria.

So, I don't think our contributor ladder is that flat. IMHO, if we keep saying that Kiali project only has two roles, anybody that wants to join to the project will quickly realize that Kiali's structure isn't that flat when trying to open PRs or starting discussions, and won't be able to figure out how to get involved properly.

If we fail to transparently document all these roles, I don't think we will be able to transparently document our workflow (which is something @xeviknal has mentioned)

**Reply by lucasponce** [reply]
> Later, you are also mentioning "issue owners", "reviewers", "QE" and "team leads". All these are also roles that people involved in Kiali are performing. We shouldn't be hiding these figures, and we should transparently provide step-in criteria.

Well, I think that one aspect is the role in a open source project (contributor, maintainer) and another aspect could be the process/flow. As an example, I don't see a "reviewer" a new role, but a task that someone performs in a specific role.

But in general +1 to document the flow better, in that sense things can improve. From the flow, it would be clearly if additional roles are necessary or as it happens, the existing roles are enough. Same for "issue owner", any contributor can be an owner, but yes, defining better the flow will help in how things should work.

**Reply by jberkus** [reply]
> If I understand correctly, the ladder lists the roles. The ladder is, simply, placing some organization and formalization to roles. So, let's not think on "the ladder" and "roles" as two separate things.

You understand correctly.

**Comment by lucasponce** [comment]
> So, my aim is to clearly identify this implicit governance and write it down to make it explicit. Let's identify current member roles, policies, processes and create one or more documents describing all this.

Also comparing with other projects with similar size, I guess we two groups that make sense: contributors and mantainers.

I agree that we should refine the implicit processes and make them explicit for those groups.

> However a consequence of an implicit governance is that, sometimes, discussions can reach a stall. IMO, this is because expectations and roles are not explicitly set.

I think we could also work in a "Technical Conflict Section" which may reflect the spirit of the project and the criteria, here there are multiple examples that may be valid like [1]

[1] https://github.com/eclipse-vertx/vert.x/blob/master/CONTRIBUTING.md#technical-conflict-process

**Comment by xeviknal** [comment]
I am completely aligned with the fact that our governance document does not reflect the whole governance of the community.

1. Document the workflow from the inception of an issue to the merge of a PR. Details like owners in each step: triage / prioritization, owners of each feature, git flow (branching, PR, approvals, merge strategy), code review styles/standards.
2. Update the list of maintainers and make it open to the public: https://github.com/orgs/kiali/teams/maintainers/members
3. I'd change the way to nominate a new maintainer: create a PR adding the person into the file. Instead of an email.
4. Code of conduct: I am fine with the content. let's update the contacts. I propose to keep pointing to multiple single email inboxes. I'd put one woman, one man and one non-binary (we don't have one in our team though, let me know if we do please) at least.
5. Release process: I'd probably refer to the workflow section where says how issues are prioritised and added into sprint. To justify what features/enhancements/bugs are included in each version.
6. Coding styles: I think it is related to code style, but a bit deeper. Maybe we are all doing good quality code but I miss more swe discussions. Talking about idiomatic go, principles, coupling. I feel that this part of the workflow is a bit weak. My feeling is that we talk more about features, ideas and technical (how k8s works for X? does Istio support Y?) discussions than code discussions. 
7. Communication: I am pretty disappointed wrt the response to people of OSM. I'd definitely prepare an easy strategy for these scenarios and reflect it in the contribution/governance file. My take is do what we did for OSM but with more certainty and less points of contacts, opinions and styles. This is pure open source and upstream first. 
> Add the topic to the next community meeting agenda for allocating a time slot for the contributors to discuss and do the first contact. If you prefer to start with a written discussion, open a discussion under the roadmap category.
8. Add a section clarifying how technical conflicts are approached and solved.

**Comment by abonas** [comment]
Are there any concrete action items from this thread to make changes to the governance?

**Reply by lucasponce** [reply]
I don't think so.
I see open topics on this thread and confusion about the goal.

**Reply by abonas** [reply]
as far as I understand, from the initial first comment the goal is to review and update Kiali's project governance which hasn't been reviewed for about half a year. Hence the outcome of this discussion should be a set of modifications applied to the current governance. @israel-hdez is my understanding correct?

**Reply by israel-hdez** [reply]
@abonas Yes, the goal is to end with updated governance docs (which is not only the governance.md file, but also others).

I was waiting for all absences/vacations to finish before proposing updates. I think starting this week, everybody in the team is available. So, I think this is a good week to start with a few "easy" changes.

**Reply by jshaughn** [reply]
Yes, @israel-hdez, I think given the above feedback it's time to introduce some proposed changes.

**Reply by dneary** [reply]
I liked @xeviknal 's comment above as an actionable framework for changes. No need to make massive changes or add a lot of administrative overhead at this point - documenting day to day operations is already a great start. Thinking about the social structure of the project at a granular level (the contributor ladder) and making it explicit is a bigger task, but it can be addressed separately, in parallel.

**Comment by nrfox** [comment]
> Contribution process: I see outdated information. Let's discuss what to update.

As has been mentioned elsewhere, explicitly defining the steps in the contribution process and who is involved would really help. For me, what has been the biggest change from other projects I have contributed to in the past is the addition of the QE role. Right now, I think it's mostly implicit that each PR needs to be assigned to and reviewed by a member of QE. Since they are a vital part of the contribution process, it would be good to define and document the QE role and maybe who all is part of the QE team so you know who to add as a reviewer for your change.

**Comment by israel-hdez** [comment]
I just opened a PR to properly document the contributor ladder: https://github.com/kiali/kiali/pull/4311
I'm hoping all people involved in Kiali project to give their :+1:  or :-1: .
